File: Java/Observer/Calculator/src/com/ericgtkb/Subject.java
package com.ericgtkb;

import java.util.HashSet;

public interface Subject {
    void addObserver(Observer obs);
    void removeObserver(Observer obs);
    void notifyObservers();
}

class DataSource implements Subject {
    private HashSet<Observer> observers;
    private int a;
    private int b;

    public DataSource() {
        observers = new HashSet<>();
    }

    public int getA() {
        return a;
    }

    public int getB() {
        return b;
    }

    public void updateData(int a, int b) {
        this.a = a;
        this.b = b;
        notifyObservers();
    }

    @Override
    public void addObserver(Observer obs) {
        observers.add(obs);
    }

    @Override
    public void removeObserver(Observer obs) {
        observers.remove(obs);
    }

    @Override
    public void notifyObservers() {
        for (var observer : observers) {
            observer.update();
        }
    }
}

File: C++/Observer/Calculator/subject.cpp
//
// Created by Eric Chen on 2019-09-05.
//

#include "subject.h"
#include "observer.h"


DataSource::DataSource() : observers{}, a{}, b{} {
}

void DataSource::add_observer(Observer& obs) {
    observers.emplace(&obs);
}

void DataSource::remove_observer(Observer& obs) {
    auto it = observers.find(&obs);
    observers.erase(it);
}

void DataSource::notify_observers() {
    for (auto observer : observers) {
        observer->update();
    }
}

int DataSource::get_a() const {
    return a;
}

int DataSource::get_b() const {
    return b;
}

void DataSource::update_data(int a, int b) {
    this->a = a;
    this->b = b;
    notify_observers();
}
